compiler/rustc_lint/src/for_loop_over_fallibles.rs

```diff
@@ -0,0 +1,188 @@
+use crate::{LateContext, LateLintPass, LintContext};
+
+use hir::{Expr, Pat};
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_infer::traits::TraitEngine;
+use rustc_infer::{infer::TyCtxtInferExt, traits::ObligationCause};
+use rustc_middle::ty::{self, List};
+use rustc_span::{sym, Span};
+use rustc_trait_selection::traits::TraitEngineExt;
+
+declare_lint! {
+    /// The `for_loop_over_fallibles` lint checks for `for` loops over `Option` or `Result` values.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// let opt = Some(1);
+    /// for x in opt { /* ... */}
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Both `Option` and `Result` implement `IntoIterator` trait, which allows using them in a `for` loop.
+    /// `for` loop over `Option` or `Result` will iterate either 0 (if the value is `None`/`Err(_)`)
+    /// or 1 time (if the value is `Some(_)`/`Ok(_)`). This is not very useful and is more clearly expressed
+    /// via `if let`.
+    ///
+    /// `for` loop can also be accidentally written with the intention to call a function multiple times,
+    /// while the function returns `Some(_)`, in these cases `while let` loop should be used instead.
+    ///
+    /// The "intended" use of `IntoIterator` implementations for `Option` and `Result` is passing them to
+    /// generic code that expects something implementing `IntoIterator`. For example using `.chain(option)`
+    /// to optionally add a value to an iterator.
+    pub FOR_LOOP_OVER_FALLIBLES,
```

Comment:
While trying to deprecate this myself, I noticed that you dropped the plural `S` from `LOOP`. I think the lint name `FOR_LOOPS_OVER_FALLIBLES` sounds better and fits the lint naming conventions better.

Reply:
It's sad that there aren't any guidelines, for me the version without the `S` sounds better. I've changed the name back to include `S`.

Reply:
There are lint naming conventions: https://rust-lang.github.io/rfcs/0344-conventions-galore.html#lints

The plural form is preferred, because it sounds better with `allow`. If `for` loop is singular, but `fallibles` is plural, it might indicate that this lint is about for loops that iterate over multiple fallibles (1:N) instead of a for loop over a fallible (1:1).